### PR TITLE
[FLINK-8802] [QS] Fix concurrent access to non-duplicated serializers.

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

This fixes the problem of multiple threads using the same serializer instance to concurrently deserialize requests.

## Brief change log

Now the `KvStateRegistry` holds a `KvStateEntry` for each registered `InternalKvState` and in that entry we have a cache that holds a copy of the serializer per thread, if the serializer is stateful. If it is stateless, we do not need to copy the serializer.

## Verifying this change

There is a test added in the `KvStateRegistryTest`, the `KvStateRegistryTest#testKvStateEntry()` and also I have tested it with an actual job (but I run it locally).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java [hunk 1]
```diff
@@ -78,13 +75,22 @@ public KvStateServerHandler(
...
 			} else {
 				byte[] serializedKeyAndNamespace = request.getSerializedKeyAndNamespace();
 
-				byte[] serializedResult = kvState.getSerializedValue(serializedKeyAndNamespace);
+				// here we remove any type check...
+				// Ideally we want to keep that the info match the state.
+				final InternalKvState state = kvState.getState();
+				final KvStateInfo info = kvState.getInfoForCurrentThread();
+
+				byte[] serializedResult = state.getSerializedValue(
```
Comment (zentol) [inline flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java]: Couldn't we synchronize on `kvState` instead of modifying all `InternalKvState` implementations?

This seems like a much safer alternative than baking in the assumption that `getSerializedValue()` can be called concurrently.

Comment (kl0u) [reply flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java]: It is not an assumption. It is accessed concurrently from the thread pool. Synchronizing on the `kvState` would serialize all accesses, which would slow down (by a lot) performance.

Comment (zentol) [reply flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java]: Ah, i mistakenly thought that `getSerializedValue()` is also called out of QS.

What would be the overhead of always duplicating serializers within `getSerializedValue()`?

Comment (kl0u) [reply flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java]: "Word on the street" :P is that `Kryo` duplication is pretty expensive. This is why I went for this solution.

Comment (zentol) [reply flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java]: well we could move the thread-caching into the `InternalKvStates`.

Comment (kl0u) [reply flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java]: True, but after also discussion with Stephan and Aljoscha we concluded that probably it is not the right place, just from a "separation-of-concerns" point of view.

That said, the more I was working on it, the more I was thinking that the `InternalKvStates` could also be a place. But this we can leave for future discussion when we re-prioritize QS.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalMergingState.java [hunk 2]
```diff
@@ -26,12 +26,14 @@
...
- * 
- * @param <N>   The type of the namespace
- * @param <IN>  The type of elements added to the state
- * @param <OUT> The type of elements 
+ *
+ * @param <K>	The type of key the state is associated to
+ * @param <N>	The type of the namespace
+ * @param <IN>	The type of elements added to the state
+ * @param <SV>	The type of elements in the state
+ * @param <OUT>	The type of elements
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalMergingState.java]: indentation is off and the description deserves an extension

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalMapState.java [hunk 3]
```diff
@@ -20,13 +20,16 @@
...
  * The peer to the {@link MapState} in the internal state type hierarchy.
  *
  * <p>See {@link InternalKvState} for a description of the internal state hierarchy.
  *
- * @param <N> The type of the namespace
- * @param <UK> Type of the values folded into the state
- * @param <UV> Type of the value in the state
+ * @param <K>	The type of key the state is associated to
+ * @param <N>	The type of the namespace
+ * @param <UK>	Type of the values folded into the state
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalMapState.java]: the description seems like a copy&paste from `FoldingState`

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalAppendingState.java [hunk 4]
```diff
@@ -24,9 +24,11 @@
...
- * 
- * @param <N>   The type of the namespace
- * @param <IN>  The type of elements added to the state
- * @param <OUT> The type of the 
+ *
+ * @param <K>	The type of key the state is associated to
+ * @param <N>	The type of the namespace
+ * @param <IN>	The type of elements added to the state
+ * @param <SV>	The type of elements in the state
+ * @param <OUT>	The type of the resulting element in the state
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalAppendingState.java]: indentation is weird

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapFoldingState.java [hunk 5]
```diff
@@ -29,20 +29,19 @@
...
+ * Heap-backed partitioned {@link FoldingState} that is snapshotted into files.
  *
- * @param <K> The type of the key.
- * @param <N> The type of the namespace.
- * @param <T> The type of the values that can be folded into the state.
- * @param <ACC> The type of the value in the folding state.
+ * @param <K>	The type of the key.
+ * @param <N>	The type of the namespace.
+ * @param <T>	The type of the values that can be folded into the state.
+ * @param <ACC>	The type of the value in the folding state.
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapFoldingState.java]: indentation is funky

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapAggregatingState.java [hunk 6]
```diff
@@ -30,18 +30,17 @@
...
- * @param <K> The type of the key.
- * @param <N> The type of the namespace.
- * @param <IN> The type of the value added to the state.
- * @param <ACC> The type of the value stored in the state (the accumulator type).
- * @param <OUT> The type of the value returned from the state.
+ * @param <K>	The type of the key.
+ * @param <N>	The type of the namespace.
+ * @param <IN>	The type of the value added to the state.
+ * @param <ACC>	The type of the value stored in the state (the accumulator type).
+ * @param <OUT>	The type of the value returned from the state.
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapAggregatingState.java]: indentation

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/AbstractHeapMergingState.java [hunk 7]
```diff
@@ -28,18 +28,19 @@
...
- * @param <K> The type of the key.
- * @param <N> The type of the namespace.
- * @param <SV> The type of the values in the state.
- * @param <S> The type of State
- * @param <SD> The type of StateDescriptor for the State S
+ * @param <K>	The type of the key.
+ * @param <N>	The type of the namespace.
+ * @param <IN>	The type of the input elements.
+ * @param <SV>	The type of the values in the state.
+ * @param <OUT>	The type of the output elements.
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/AbstractHeapMergingState.java]: indentation

File: flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateRegistry.java [hunk 8]
```diff
@@ -136,13 +138,13 @@ public void unregisterKvState(
 	}
 
 	/**
-	 * Returns the KvState instance identified by the given KvStateID or
-	 * <code>null</code> if none is registered.
+	 * Returns the KvState instance identified by the given KvStateID along with
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateRegistry.java]: would rephrase to `KvStateEntry` containing the `KvState`

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java [hunk 9]
```diff
@@ -686,7 +686,7 @@ public void testClientServerIntegration() throws Throwable {
 				state.update(201 + i);
 
 				// we know it must be a KvStat but this is not exposed to the user via State
```
Comment (zentol) [inline flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/ClientTest.java]: typo: KvStat

File: flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java [hunk 10]
```diff
@@ -78,13 +75,22 @@ public KvStateServerHandler(
...
-			final InternalKvState<?> kvState = registry.getKvState(request.getKvStateId());
+			final KvStateEntry<?, ?, ?> kvState = registry.getKvState(request.getKvStateId());
 			if (kvState == null) {
 				responseFuture.completeExceptionally(new UnknownKvStateIdException(getServerName(), request.getKvStateId()));
 			} else {
 				byte[] serializedKeyAndNamespace = request.getSerializedKeyAndNamespace();
 
-				byte[] serializedResult = kvState.getSerializedValue(serializedKeyAndNamespace);
+				// here we remove any type check...
+				// Ideally we want to keep that the info match the state.
```
Comment (zentol) [inline flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java]: you can retain type safety:

Call from the handler:
```
byte[] serializedResult = getSerializedValue(kvState, serializedKeyAndNamespace);
```

Added method:
```
private static <K, N, V> byte[] getSerializedValue(KvStateEntry<K, N, V> entry, byte[] serializedKeyAndNamespace) throws Exception {
		InternalKvState<K, N, V> state = entry.getState();
		KvStateInfo<K, N, V> infoForCurrentThread = entry.getInfoForCurrentThread();
		
		return state.getSerializedValue(
			serializedKeyAndNamespace,
			infoForCurrentThread.getKeySerializer(),
			infoForCurrentThread.getNamespaceSerializer(),
			infoForCurrentThread.getStateValueSerializer()
		);
	}
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateEntry.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		this.state = Preconditions.checkNotNull(state);
+		this.stateInfo = new KvStateInfo<>(
+				state.getKeySerializer(),
+				state.getNamespaceSerializer(),
+				state.getValueSerializer()
+		);
+
+		this.serializerCache =
+				stateInfo.duplicate() == stateInfo
+						? null							// if the serializers are stateless, we do not need a cache
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateEntry.java]: An empty map would be acceptable here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateInfo.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+	 * <p>This method assumes correct implementation of the {@link TypeSerializer#duplicate()}
+	 * method of the included serializers.
+	 */
+	public KvStateInfo<K, N, V> duplicate() {
+		final TypeSerializer<K> dupKeySerializer = keySerializer.duplicate();
+		final TypeSerializer<N> dupNamespaceSerializer = namespaceSerializer.duplicate();
+		final TypeSerializer<V> dupSVSerializer = stateValueSerializer.duplicate();
+
+		if (
+				dupKeySerializer == keySerializer &&
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateInfo.java]: odd formatting

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java [hunk 13]
```diff
@@ -140,19 +159,28 @@ public boolean contains(UK userKey) {
...
+
+		Map<UK, UV> result = stateTable.get(keyAndNamespace.f0, keyAndNamespace.f1);
+
+		if (result == null) {
 			return null;
 		}
 
-		TypeSerializer<UK> userKeySerializer = stateDesc.getKeySerializer();
-		TypeSerializer<UV> userValueSerializer = stateDesc.getValueSerializer();
+		final HashMapSerializer<UK, UV> serializer = (HashMapSerializer<UK, UV>) valueSerializer;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java]: this shouldn't be necessary. why can't we just pass the map serialzer into `serializeMap`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java [hunk 14]
```diff
@@ -40,6 +54,81 @@
...
+		for (KvStateInfo<?, ?, ?> infoA: infos) {
+			boolean found = false;
+			for (KvStateInfo<?, ?, ?> infoB: infos) {
+				if (infoA == infoB) {
+					if (found) {
+						Assert.fail("Already found");
+					}
+					found = true;
+				} else {
+					Assert.assertTrue(infoA != infoB && infoA.equals(infoB));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java]: `infoA != infoB` is redundant

File: flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java [hunk 15]
```diff
@@ -40,6 +54,81 @@
...
+		Assert.assertEquals(threads, kvState.getCacheSize());
+
+		latch2.countDown();
+
+		for (KvStateInfo<?, ?, ?> infoA: infos) {
+			boolean found = false;
+			for (KvStateInfo<?, ?, ?> infoB: infos) {
+				if (infoA == infoB) {
+					if (found) {
+						Assert.fail("Already found");
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java]: needs a better error message

File: flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java [hunk 16]
```diff
@@ -40,6 +54,81 @@
...
+		}
+
+		kvStateRegistry.unregisterKvState(
+				jobID,
+				jobVertexId,
+				keyGroupRange,
+				registrationName,
+				stateID);
+
+		// we have to call for garbage collection to be sure that everything is cleared up.
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java]: ?

File: flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java [hunk 17]
```diff
@@ -40,6 +54,81 @@
...
+			new Thread(() -> {
+				final KvStateEntry<?, ?, ?> kvState = kvStateRegistry.getKvState(stateID);
+				final KvStateInfo<?, ?, ?> stateInfo = kvState.getInfoForCurrentThread();
+				infos.add(stateInfo);
+
+				latch1.countDown();
+				try {
+					latch2.await();
+				} catch (InterruptedException e) {
+					Assert.fail(e.getMessage());
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java]: this would never directly fail the test and we may loose the exception

File: flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java [hunk 18]
```diff
@@ -40,6 +54,81 @@
...
+		Assert.assertEquals(threads, infos.size());
+		Assert.assertEquals(threads, kvState.getCacheSize());
+
+		latch2.countDown();
+
+		for (KvStateInfo<?, ?, ?> infoA: infos) {
+			boolean found = false;
+			for (KvStateInfo<?, ?, ?> infoB: infos) {
+				if (infoA == infoB) {
+					if (found) {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java]: `found` needs a better name or a comment what it means

File: flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java (hunk 16 above)
Comment (kl0u) [reply flink-runtime/src/test/java/org/apache/flink/runtime/query/KvStateRegistryTest.java]: Forgotten from previous experiments that were using a weak hash map. I will remove it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java (hunk 13 above)
Comment (kl0u) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java]: The problem is that the Heap Backend uses the `HashMapSerializer` while RocksDB uses `MapSerializer`. So here it should be sth like `TypeSerializer<? extends Map<UK, UV>>` that then should somehow be casted to the appropriate serializer somehow in order to get the key and the value serializer. The solution that we have here is a bit cleaner, although the best would be to remove the `HashMapSerializer` and have a `MapSerializer` in here.

**Comment by kl0u** [comment]
Thanks for the review @zentol ! I updated the PR integrating your comments, or commenting on why I did not do it. Please have a look and let me know if you have more comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateEntry.java [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	public KvStateInfo<K, N, V> getInfoForCurrentThread() {
+		return isSerializerStateless
+				? stateInfo
+				: serializerCache.computeIfAbsent(Thread.currentThread(), t -> stateInfo.duplicate());
+	}
+
+	public void clear() {
+		if (serializerCache != null) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateEntry.java]: unnecessary null check

File: flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateEntry.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+			serializerCache.clear();
+		}
+	}
+
+	@VisibleForTesting
+	public int getCacheSize() {
+		return serializerCache == null ? -1 : serializerCache.size();
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateEntry.java]: unnecessary null check

File: flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateEntry.java [hunk 21]
```diff
@@ -39,29 +39,27 @@
...
 				state.getNamespaceSerializer(),
 				state.getValueSerializer()
 		);
-
-		this.serializerCache =
-				stateInfo.duplicate() == stateInfo
-						? null							// if the serializers are stateless, we do not need a cache
-						: new ConcurrentHashMap<>();
+		this.serializerCache = new ConcurrentHashMap<>();
+		this.isSerializerStateless = stateInfo.duplicate() == stateInfo;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/query/KvStateEntry.java]: -> areSerializersStateless?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java (hunk 13 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java]: i mean why isn't the signature of `KvStateSerializer.serializeMap`:
```
KvStateSerializer.serializeMap(Map<UK, UV> map, TypeSerializer<Map<UK, UV>> serializer);
```

with the following implementation:
```
...
serializeMap(Map<UK, UV> map, TypeSerializer<Map<UK, UV>> serializer) {
		if (map != null) {
			DataOutputSerializer dos = new DataOutputSerializer(32);
			serializer.serialize(map, dos);
			return dos.getCopyOfBuffer();
		} else {
			return null;
		}
```

Why deal with the map key/value entries at all outside the serializer?

Comment (kl0u) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java]: The reason is that RocksDB returns an iterator that gets lazily populated as you call next() while the serialize() of the MapSerializer expects a Map. If it were to go with your option, we would have to iterate over the map twice, once to create the map, and then to serialize it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalKvState.java [hunk 22]
```diff
@@ -70,10 +88,18 @@
...
+	 * stateful (e.g. serializers) should be either duplicated or protected from undesired
+	 * consequences of concurrent invocations.
+	 *
 	 * @param serializedKeyAndNamespace Serialized key and namespace
 	 * @return Serialized value or <code>null</code> if no value is associated with the key and namespace.
 	 * 
 	 * @throws Exception Exceptions during serialization are forwarded
 	 */
-	byte[] getSerializedValue(byte[] serializedKeyAndNamespace) throws Exception;
+	byte[] getSerializedValue(
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalKvState.java]: I would suggest to store the serializer in a thread local variable. The current solution is a bit confusing because this interface suddenly exposes serializers and caller have to provide serialzer in the `getSerializedValue` method. In my opinion this interface does not make much sense in this way. Furthermore, the serializers are copied externally into something that looks like a custom-build thread local. I suggest having the serializers thread local in the base class and bringing this interface back to the original form. There is also only one threadpool, dedicated for queryable state that would hold the serializers and even the current solution has a dedicated cleanup method. In that place, we can just clean the thread locals.

**Comment by kl0u** [comment]
@StefanRRichter and @zentol  I will ping you when I am done with the refactoring.

**Comment by kl0u** [comment]
This is ready for another review @StefanRRichter and @zentol

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalQueryableKvState.java [hunk 23]
```diff
@@ -16,46 +16,42 @@
...
-public class KvStateEntry<K, N, V> {
+public abstract class InternalQueryableKvState<K, N, V> implements InternalKvState<N> {
 
-	private final InternalKvState<K, N, V> state;
 	private final KvStateInfo<K, N, V> stateInfo;
-
 	private final boolean areSerializersStateless;
 
-	private final ConcurrentMap<Thread, KvStateInfo<K, N, V>> serializerCache;
+	private final ConcurrentMap<Thread, KvStateInfo<K, N, V>> serializerCache = new ConcurrentHashMap<>(4);
```
Comment (sihuazhou) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/internal/InternalQueryableKvState.java]: nit:  just wonder why didn't use ThreadLocal<KvStateInfo<K, N, V>> provided by JDK...

**Comment by kl0u** [comment]
That is because cleaning up the cache is happening from another thread, other than the one 
accessing the serializers, and ThreadLocal does not have a clear() method that you can call
from another thread and clean all the state in it. Each thread can only clean its own state.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/AbstractHeapState.java [hunk 24]
```diff
@@ -31,14 +31,13 @@
  * Base class for partitioned {@link State} implementations that are backed by a regular
  * heap hash map. The concrete implementations define how the state is checkpointed.
  *
- * @param <K> The type of the key.
- * @param <N> The type of the namespace.
- * @param <SV> The type of the values in the state.
- * @param <S> The type of State
- * @param <SD> The type of StateDescriptor for the State S
+ * @param <K>	The type of the key.
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/AbstractHeapState.java]: All of the generic parameters seem to have a tab in them. That's why they look wonky here. Maybe should fix those.
